common/cuda_hip/matrix/bccoo_kernels.hpp.inc

```diff
@@ -0,0 +1,621 @@
+LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+******************************<GINKGO LICENSE>*******************************/
+
+#include "common/cuda_hip/matrix/bccoo_helper.hpp.inc"
+#include "core/matrix/bccoo_aux_structs.hpp"
+
+namespace {
+
+// #define OLD_BLOCK 1
+
+
+/**
+ * The device function of BCCOO spmv
+ *
+ * @param nnz  the number of nonzeros in the matrix
+ * @param num_lines  the maximum round of each warp
+ * @param val  the value array of the matrix
+ * @param col  the column index array of the matrix
+ * @param row  the row index array of the matrix
+ * @param b  the input dense vector
+ * @param b_stride  the stride of the input dense vector
+ * @param c  the output dense vector
+ * @param c_stride  the stride of the output dense vector
+ * @param scale  the function on the added value
+ *
+ * @tparam ValueType  type of values stored in the matrix
+ * @tparam IndexType  type of matrix indexes stored in the structure
+ * @tparam Closure  type of the function used to write the result
+ */
+template <int subwarp_size = config::warp_size, typename ValueType,
+          typename IndexType, typename Closure>
+__device__ void spmv_kernel(
+    const size_type nnz, const size_type num_blks, const size_type block_size,
+    const size_type num_lines, const uint8* __restrict__ chunk_data,
+    //                            const IndexType* __restrict__ offsets_data,
+    const size_type* __restrict__ offsets_data,
+    const uint8* __restrict__ types_data,
+    const IndexType* __restrict__ cols_data,
+    const IndexType* __restrict__ rows_data, const ValueType* __restrict__ b,
+    const size_type b_stride, ValueType* __restrict__ c,
+    const size_type c_stride, Closure scale)
+{
+    const auto column_id = blockIdx.y;
+    const auto start_blk = blockIdx.x;
+    const auto jump_blk = gridDim.x;
+
+    const auto start_in_blk = threadIdx.y * subwarp_size + threadIdx.x;
+    const auto jump_in_blk = blockDim.y * subwarp_size;
+
+    ValueType temp_val = zero<ValueType>();
+    bool new_value = false;
+
+    for (IndexType blk = start_blk; blk < num_blks; blk += jump_blk) {
+        const auto tile_block =
+            group::tiled_partition<subwarp_size>(group::this_thread_block());
+
+        size_type block_size_local =
+            std::min(block_size, nnz - block_size * blk);
+        compr_idxs idxs = {};
+        compr_blk_idxs blk_idxs = {};
+        idxs.blk = blk;
+        idxs.shf = offsets_data[blk];
+        idxs.row = rows_data[blk];
+        init_block_indices(rows_data, cols_data, block_size_local, idxs,
+                           types_data[blk], blk_idxs);
+#ifdef OLD_BLOCK
```

Comment:
If it is not an option such that users can use it, you also need to delete it and the corresponding code.

Reply:
Done!!